extra_proposal/proposal.py

```diff
@@ -0,0 +1,463 @@
+
+        times = dict()
+        if "Run end" in events and "Run begin" in events:
+            times["Run"] = events["Run end"] - events["Run begin"]
+        if "Migration end" in events and "Migration begin" in events:
+            times["Migration"] = events["Migration end"] - events["Migration begin"]
+        if "Cal end" in events and "Cal begin" in events:
+            key = "Calibration" if cal_requests == 1 else f"Calibration attempt {cal_requests}"
+            times[key] = events["Cal end"] - events["Cal begin"]
+
+        fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(9, 5))
+        for (event, x) in events.items():
+            if not (event.startswith("Cal") and cal_requests > 1):
+                ax1.scatter(x, 1, label=event)
+
+        ax1.legend(bbox_to_anchor=(1.01, 1.05))
+        ax1.tick_params(left=False, labelleft=False)
+        ax1.set_title("Timeline of run events")
+        ax1.grid(axis="x")
+
+        ax2.bar(times.keys(), [x.total_seconds() / 60 for x in times.values()])
+        ax2.set_ylabel("Time [minutes]")
+        ax2.set_title("Time taken in each state")
+        ax2.grid(axis="y")
+
+        fig.suptitle(f"p{self.proposal}, r{self.run_num} - {self.run_type()} - {self.sample_name()}")
+
+        fig.tight_layout()
+
+        return ax1
+
+def _cache_by_run(func):
+    @wraps(func)
+    def wrapper(self: 'Proposal', run):
+        key = (run, func.__name__)
+        if self._enable_cache and key in self._cached_data:
+            return self._cached_data[key]
+
+        value = func(self, run)
+        if self._enable_cache:
+            self._cached_data[key] = value
+
+        return value
+
+    return wrapper
+
+
+class Proposal:
+    def __init__(
+        self,
+        proposal: Optional[int | str] = None,
+        user_id: Optional[str] = None,
+        user_secret: Optional[str] = None,
+        user_email: Optional[str] = None,
+        timeout=10,
+        enable_cache=True,
+    ):
+        """Proposal object.
+        It can be instantiated as:
+        1. proposal = Proposal(), if in a proposal directory.
+        2. proposal = Proposal(2112)
+
+        Args:
+            proposal (Optional[int  |  str], optional): Proposal number. Defaults to None.
+            user_id (Optional[str], optional): UID (can be generated at https://in.xfel.eu/metadata/oauth/applications). Defaults to None.
+            user_secret (Optional[str], optional): Secret (can be generated at https://in.xfel.eu/metadata/oauth/applications). Defaults to None.
+            user_email (Optional[str], optional): User's email. Defaults to None.
+
+        Raises:
+            ProposalNotFoundError: The proposal does not exist.
+        """
+
+        if proposal is None:
+            # are we in a proposal folder?
+            cwd = os.getcwd()
+
+            if cwd.startswith(str(DATA_ROOT_DIR)):
+                try:
+                    proposal = cwd.replace(str(DATA_ROOT_DIR), "").split(
```

Comment:
One slight complication with using PWD is that the usual paths are symlinks and it's possible to end up in the absolute path instead, like `/gpfs/exfel/u/usr/MID/202401/p004977` instead of `/gpfs/exfel/exp/MID/202401/p004977`. Could we handle those cases too? Anecdotally this has only ever happened to me in the `usr/` and `scratch/` directories so I think we only need to handle those for now.

Reply:
I wanted to discuss whether we keep the use-the-cwd logic from Luca's code at all. I can see the convenience, but to my mind it's a bit too much magic - I think it's confusing if you copy & paste code to a different directory and it stops working or does something different.

If other people like that, a compromise could be that we recognise any path under a proposal, including resolving relative paths, so passing `"."` to mean the CWD could work. But I'm inclined to require a more explicit specification of the proposal.

Reply:
Mmm I see. I'd be ok with removing it then, or maybe put it behind e.g. `Proposal.from_pwd()`.